llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -627,6 +661,57 @@ static string_vector saveResultSymbolsLists(string_vector &ResSymbolsLists,
     }                                                                          \
   }
 
+// Helper function for creating Inliner pass.
+// The approach is taken from opt tool.
+static Pass *createFunctionInliningPassHelper() {
+  if (OptLevelO0)
+    return createFunctionInliningPass(0, 0, false);
+
+  if (OptLevelO1)
+    return createFunctionInliningPass(1, 0, false);
+
+  if (OptLevelO2)
+    return createFunctionInliningPass(2, 0, false);
+
+  if (OptLevelOs)
+    return createFunctionInliningPass(2, 1, false);
+
+  if (OptLevelOz)
+    return createFunctionInliningPass(2, 2, false);
+
+  if (OptLevelO3)
+    return createFunctionInliningPass(3, 0, false);
+
+  return createFunctionInliningPass();
+}
+
+// When ESIMD code was separated from the regular SYCL code,
+// we can safely process ESIMD part.
```

Comment:
This is probably an architecture question/concern: why can't we leave a single optimization pipeline which is executed before `sycl-post-link` invocation?

I understand that IGC VC backend most likely expects LLVM IR in some particular form to do its job right, but can't it be improved? It seems a bit strange to me that we have to maintain two optimization pipelines in two different places. Moreover, I guess that the first pipeline before `sycl-post-link` will still be executed to optimize regular non-ESIMD code, right?

Note: I'm okay if we have to launch a few more passes in order to launch some `ESIMD`-specific ones, but I would like to have a single optimization pipeline (potentially customized compared to the upstream one) for general-purpose optimizations.

Reply:
@AlexeySachkov, yes, those are ESIMD-specific transformations that should be applied to ESIMD code only, otherwise, it will break regular SYCL code. So, it should be done after splitting. Or we should clone the code that is shared between ESIMD and regular SYCL kernels and then do the ESIMD lowering. Splitting first and then transform, eliminates the need to clone shared code, so I guess this is definitely an advantage.

Yes, we have general-purposes clean-up passes after ESIMD lowering. So, yes, we have 2 sets of general-purpose passes running on ESIMD code. I believe we have to have the second set after ESIMD lowering. @cmc-rep can confirm.

Reply:
> those are ESIMD-specific transformations that should be applied to ESIMD code only, otherwise, it will break regular SYCL code.

I'm not against ESIMD-specific transformation, I'm concerned about additional optimization passes we launch in `sycl-post-link`

> Yes, we have general-purposes clean-up passes after ESIMD lowering. So, yes, we have 2 sets of general-purpose passes running on ESIMD code. I believe we have to have the second set after ESIMD lowering. @cmc-rep can confirm.

To me it seems better if those post-processing is moved into vc backend as pre-processing passes, because this is a thing specific to a single backend, not even to a language constructs.